### PR TITLE
Ft user role 181414757

### PR DESCRIPTION
**Pull request by itisWasp**
### What does this PR do?
Assign Role to a user by their email
### Description of Task to be completed?
- assign roles to any staff member (via their email address)
 - The default role is requester when a user is registered
 - logged in as a super administrator, I should be able to assign roles to users by using their email addresses.`
 - Creating a super admin when the server starts

### How should this be manually tested?
Get started By Running commands bellow:
  - clone repo
  - switch to Ftuserrole181414757 branch
  - ``yarn install``: to install all dependencies
Test if you can access data stored in your database:
  - Create .env file
  - add a database URL using "TEST_DB" as variable
  - ``yarn migrate --env test``: to migrate table
  - ``yarn test``: to run the test suite

### What are the relevant pivotal tracker stories?
[#181414757](https://www.pivotaltracker.com/story/show/181414757)
### Questions:

### CODE REVIEW

File: src/config/init.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+      });
+
+      if (!admin) {
+        createUserService({
+          firstName: config.adminAccountFirstname,
+          lastName: config.adminAccountLastname,
+          email: config.adminAccountEmail,
+          userName: config.adminAccountUsername,
+          password: hashedPassword,
+          roleId: 4,
```
Comment (ericus123) [inline src/config/init.js]: This id should not be hardcoded

File: src/app.js [hunk 2]
```diff
@@ -1,32 +1,23 @@
...
-  res.header(
-    'Access-Control-Allow-Headers',
-    'Origin, X-Requested-With, Content-Type, Accept'
-  );
-  next();
-});
-app.use(express.json());
-app.use(looger);
-
-app.use('/api', routes);
```
Comment (didasmbalanya) [inline src/app.js]: Why did you remove the following?
- Cors
- express.json body parser?
- logger
- root route?

File: src/app.js [hunk 3]
```diff
@@ -1,32 +1,23 @@
...
-
+app.use('/api/auth', authRoute);
+app.use('/api/roles', rolesRoute);
+app.use('/api/docs', docs);
```
Comment (didasmbalanya) [inline src/app.js]: all this should be done in routes/index.js and then only the routes object is passed inside app.js, as it was previously

File: src/config/config.js [hunk 4]
```diff
@@ -0,0 +1,11 @@
+export default {
+  db: process.env.DB_CONNECT,
+  // set this to build
+  // dir: 'build/',
+  sessionSecret: process.env.SESSION_SECRET || 'MEAN',
+  adminAccountFirstname: process.env.ADMIN_ACCOUNT_FIRSTNAME || 'Admin',
+  adminAccountLastname: process.env.ADMIN_ACCOUNT_LASTNAME || 'User',
+  adminAccountUsername: process.env.ADMIN_ACCOUNT_USERNAME || 'Admin',
+  adminAccountEmail: process.env.ADMIN_ACCOUNT_EMAIL || 'admin@admin.com',
+  adminAccountPassword: process.env.ADMIN_ACCOUNT_PASSWORD || 'password',
```
Comment (ericus123) [inline src/config/config.js]: Don't reveal Admin credentials in any way.
They should only be hidden

File: src/controllers/user.js [hunk 5]
```diff
@@ -54,5 +54,30 @@ class UserController {
...
+
+  static assignRoles = async (req, res) => {
+    const userRole = req.body.RoleId;
+    const userExist = await User.findOne({
+      where: { email: req.body.Email },
+      include: [{ model: Role, attributes: ['id', 'name'] }],
+    });
+
+    if (!userExist) {
+      return res.status(400).json({ message: 'User does not exist' });
```
Comment (didasmbalanya) [inline src/controllers/user.js]: error status 404 (not found)

File: src/validations/UserValidation.js [hunk 6]
```diff
@@ -71,6 +72,26 @@ class UserValidation {
       });
     }
   };
+
+  static verifyRoles = (req, res, next) => {
+    try {
+      const value = roleEntryValidation.validate(req.body);
+      if (value.error) {
+        res.status(400).json({
```
Comment (didasmbalanya) [inline src/validations/UserValidation.js]: 422 - validation errors

File: src/middlewares/protected.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+        .status(401)
+        .json({ message: 'Token Not Provided Please Check Headers' });
+    }
+
+    try {
+      const verified = jwt.verify(token, process.env.TOKEN_SECRET);
+      const { roleId } = verified;
+
+      console.log(roleId);
+
```
Comment (ericus123) [inline src/middlewares/protected.js]: remove this

File: src/helpers/CheckUser.js [hunk 8]
```diff
@@ -1,8 +1,21 @@
...
+export const roleEntryValidation = Joi.object({
+  RoleId: Joi.string()
+    .valid('1', '2', '3')
+    .required()
+    .pattern(/^[0-9]/)
+    .messages({
+      'string.pattern.base': '{{#label}} must contain a valid id',
+      'string.valid.base':
+        '{{#label}} must be written as requester : 1, manager : 2, admin : 3',
+    }),
```
Comment (didasmbalanya) [inline src/helpers/CheckUser.js]: isn't this supposed to be a number?

**Review by ericus123 — CHANGES_REQUESTED** [review]

File: src/database/models/role.js [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+  Role.init(
+    {
+      name: DataTypes.STRING,
+    },
+    {
+      sequelize,
+      modelName: 'Role',
+    }
+  );
+  return Role;
```
Comment (didasmbalanya) [inline src/database/models/role.js]: role model is not the same as the migration, so this might misguide. There is no primary key and the role name is not even required. Keep the model consistent with the migration so that the actual table and the model are consitent

File: src/controllers/user.js [hunk 10]
```diff
@@ -54,5 +54,30 @@ class UserController {
...
+    }
+
+    try {
+      await User.update(
+        { roleId: userRole },
```
Comment (didasmbalanya) [inline src/controllers/user.js]: make sure userRole is a number

File: src/database/config/config.js [hunk 11]
```diff
@@ -2,13 +2,31 @@ require('dotenv').config();
 
 export const development = {
   url: process.env.DEV_DB,
+  dialectOptions: {
+    ssl: {
+      require: true,
+      rejectUnauthorized: false,
+    },
+  },
```
Comment (didasmbalanya) [inline src/database/config/config.js]: repetition, already implemented in src/database/config/database.js

File: src/database/config/config.js [hunk 12]
```diff
@@ -2,13 +2,31 @@ require('dotenv').config();
...
   dialect: 'postgres',
 };
 export const test = {
   url: process.env.TEST_DB,
+  dialectOptions: {
+    ssl: {
+      require: true,
+      rejectUnauthorized: false,
+    },
+  },
```
Comment (didasmbalanya) [inline src/database/config/config.js]: repetition, already implemented in src/database/config/database.js

File: src/database/config/config.js [hunk 13]
```diff
@@ -2,13 +2,31 @@ require('dotenv').config();
...
   dialect: 'postgres',
 };
 export const production = {
   url: process.env.DB_CONNECT,
+  dialectOptions: {
+    ssl: {
+      require: true,
+      rejectUnauthorized: false,
+    },
+  },
```
Comment (didasmbalanya) [inline src/database/config/config.js]: repetition, already implemented in src/database/config/database.js

File: src/database/migrations/20220405214034-create-role.js [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+  async up(queryInterface, Sequelize) {
+    await queryInterface.createTable('Roles', {
+      id: {
+        allowNull: false,
+        autoIncrement: true,
+        primaryKey: true,
+        type: Sequelize.INTEGER,
+      },
+      name: {
+        type: Sequelize.STRING,
```
Comment (didasmbalanya) [inline src/database/migrations/20220405214034-create-role.js]: Should be an enum type, with only a list of allowed role names

File: src/database/migrations/20220405214715-modify_users_add_new_fields.js [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+        allowNull: true,
+        type: Sequelize.INTEGER,
+        onDelete: 'CASCADE',
+        references: {
+          model: 'Roles',
+          key: 'id',
+        },
+      }),
+    ]);
+  },
```
Comment (didasmbalanya) [inline src/database/migrations/20220405214715-modify_users_add_new_fields.js:14]: good work here

File: src/database/models/user.js [hunk 16]
```diff
@@ -9,6 +9,10 @@ export default (sequelize, DataTypes) => {
      */
     static associate(models) {
       // define association here
+      User.hasOne(models.Role, {
+        foreignKey: 'id',
+        onDelete: 'CASCADE',
+      });
```
Comment (didasmbalanya) [inline src/database/models/user.js:10]: good job here, just revert to es6 though

File: src/database/seeders/20220302143415-demo-user.js [hunk 17]
```diff
@@ -1,32 +1,25 @@
...
-        password:
-          '$2b$10$bQUIGi5mWZ8NRFMN3.VjgOiV1fPTYExzxwhOpL.m/bgD6tiLS2hxm',
-        firstName: 'John',
-        lastName: 'Doe',
+        name: roles.default.ADMIN,
+        createdAt: new Date(),
+        updatedAt: new Date(),
+      },
+      {
+        name: roles.default.SUPER_ADMIN,
```
Comment (didasmbalanya) [inline src/database/seeders/20220302143415-demo-user.js]: do not update existing seeds just as models, create new ones and do the changes to pre-existing data if you have to

File: src/database/seeders/20220311101827-demo-user.js [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+
+  async down(queryInterface, Sequelize) {
+    /**
+     * Add commands to revert seed here.
+     *
+     * Example:
+     * await queryInterface.bulkDelete('People', null, {});
+     */
+  },
+};
```
Comment (didasmbalanya) [inline src/database/seeders/20220311101827-demo-user.js]: is this meant to do something?

File: src/helpers/CheckUser.js [hunk 19]
```diff
@@ -1,8 +1,21 @@
+import Joi from 'joi';
 import models from '../database/models';
 
 const { User } = models;
 
-const checkUser = async (params) => {
+export const roleEntryValidation = Joi.object({
+  RoleId: Joi.string()
+    .valid('1', '2', '3')
```
Comment (didasmbalanya) [inline src/helpers/CheckUser.js:9]: is this supposed to be strings or numbers?

File: src/helpers/roles.js [hunk 20]
```diff
@@ -0,0 +1,7 @@
+const roles = {
+  ADMIN: 'admin',
+  REQUESTER: 'requester',
+  MANAGER: 'manager',
+  SUPER_ADMIN: 'super_admin',
+};
+export default roles;
```
Comment (didasmbalanya) [inline src/helpers/roles.js]: Even on the database level the roles should be one of this. hence "Enums"

File: src/middlewares/protected.js [hunk 21]
```diff
@@ -0,0 +1,32 @@
+import jwt from 'jsonwebtoken';
+
+class ProtectedRoute {
+  static verifyToken = async (req, res, next) => {
```
Comment (didasmbalanya) [inline src/middlewares/protected.js]: a better name would be "superadminAuth" since it checks if the user is a super admin rather than if they are verified

File: src/services/RolesServices.js [hunk 22]
```diff
@@ -0,0 +1,43 @@
+import { User } from '../database/models/user';
+
+import { Role } from '../database/models/role';
+
+export const createOne = (data) => {
+  const created = models.Role.create(data);
```
Comment (didasmbalanya) [inline src/services/RolesServices.js]: async await

File: src/services/RolesServices.js [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+  return created;
+};
+
+export const findRole = (data) => {
+  const found = Role.findOne({ where: { name: data.name } });
+  return found;
+};
+export const findRoles = (data) => {
```
Comment (didasmbalanya) [inline src/services/RolesServices.js]: Better naming? FindRolesAndCount ?

File: src/services/RolesServices.js [hunk 24]
```diff
@@ -0,0 +1,43 @@
...
+    { where: { roleId: query.role_id } }
+  );
+  return changes;
+};
+export const deleteOne = (data) => {
+  const deleted = Role.destroy({ where: { id: data } });
+  return deleted;
+};
+
+export const roles = async () => {
```
Comment (didasmbalanya) [inline src/services/RolesServices.js]: findAllRoles

File: test/roles.test.js [hunk 25]
```diff
@@ -0,0 +1,60 @@
...
+    });
+
+    before(async () => {
+      const res = await request(app).post('/api/auth/login').send({
+        email: `admin@admin.com`,
+        password: 'password',
+      });
+      console.log(res.body);
+      superToken = res.body.data.token;
+    });
```
Comment (didasmbalanya) [inline test/roles.test.js]: why not do all this is the same before hook?

**Review by didasmbalanya — CHANGES_REQUESTED** [review]

File: src/controllers/user.js [hunk 26]
```diff
@@ -54,5 +54,30 @@ class UserController {
       res.status(400).json({ error: error.message });
     }
   };
+
+  static assignRoles = async (req, res) => {
+    const userRole = req.body.RoleId;
+    const userExist = await User.findOne({
+      where: { email: req.body.Email },
```
Comment (Robertishimwe) [inline src/controllers/user.js]: remember if someone signed up with Facebook he does not have email. how will admin assign role to this user? i think you can use id instead of email

Comment (itisWasp) [reply src/controllers/user.js]: > remember if someone signed up with Facebook he does not have email. how will admin assign role to this user? i think you can use id instead of email

@Robertishimwe The Task says that the role is suppossed to be assigned to the user using an email so I think there is a separate one for that me I am just following the task description.

Comment (Robertishimwe) [reply src/controllers/user.js]: Ok

Comment (didasmbalanya) [reply src/controllers/user.js]: Please do as @Robertishimwe suggests.

File: src/database/seeders/20220302143415-demo-user.js (hunk 17 above)
Comment (niyojuste) [reply src/database/seeders/20220302143415-demo-user.js]: @itisWasp I wasn't able to tag the line but, check this file, the down function has issues.
- it is outside the module exports and uses es6 exports
- it is unseeding for 'Users' model when the seeding is for 'Roles' model

File: src/database/seeders/20220311101827-demo-user.js (hunk 18 above)
Comment (niyojuste) [reply src/database/seeders/20220311101827-demo-user.js]: @itisWasp Perhaps you could use this to seed and unseed 'Users'

File: package.json [hunk 27]
```diff
@@ -14,10 +14,10 @@
     "unseed": "yarn sequelize-cli db:seed:undo:all",
     "build": "babel src -d lib",
     "start": "yarn run build && cross-env NODE_ENV=production node lib/index.js",
-    "dev": "cross-env NODE_ENV=development nodemon --exec babel-node -r dotenv/config src/index.js",
+    "dev": " cross-env NODE_ENV=development yarn migrate && cross-env NODE_ENV=development yarn seed && cross-env NODE_ENV=development nodemon --exec babel-node -r dotenv/config src/index.js",
```
Comment (Procech20) [inline package.json]: Why do you have to change the environment in the script. You can add the `NODE_ENV` in the `.env` file and define your environment there instead of doing it in the script

File: src/config/init.js [hunk 28]
```diff
@@ -0,0 +1,38 @@
+import bcrypt from 'bcrypt';
+import config from './config';
+import models from '../database/models';
+import createUserService from '../services/UserServices';
+
+const { User } = models;
+
+export default () => {
+  return new Promise((resolve, reject) => {
```
Comment (Procech20) [inline src/config/init.js]: We are currently using try...catch, try to use them instead of resolve, reject

File: src/controllers/user.js [hunk 29]
```diff
@@ -54,5 +55,40 @@ class UserController {
...
+      res.status(201).json(users);
+    } catch (error) {
+      res.status(400).json({ message: error.message });
+    }
+  };
+
+  static assignRoles = async (req, res) => {
+    const userRole = req.body.RoleId;
+    const userId = req.params.id;
+    const userExist = await User.findOne({
```
Comment (Procech20) [inline src/controllers/user.js]: This is supposed to be done in the services and you call the service here

File: src/database/models/role.js [hunk 30]
```diff
@@ -0,0 +1,27 @@
+const { Model } = require('sequelize');
+
+module.exports = (sequelize, DataTypes) => {
+  class Role extends Model {
+    /**
+     * Helper method for defining associations.
```
Comment (Procech20) [inline src/database/models/role.js]: These comments cause linting errors and they don't really do anything or guide anyone. Please remove them

File: src/middlewares/superadminAuth.js [hunk 31]
```diff
@@ -0,0 +1,30 @@
...
+        .status(401)
+        .json({ message: 'Token Not Provided Please Check Headers' });
+    }
+
+    try {
+      const verified = jwt.verify(token, process.env.TOKEN_SECRET);
+      const { roleId } = verified;
+
+      if (roleId !== 4) {
+        return res.status(401).json({
```
Comment (Procech20) [inline src/middlewares/superadminAuth.js]: The status should be '403' which is related to Authorization.

**Review by Procech20 — CHANGES_REQUESTED** [review]

File: src/database/migrations/20220405214034-create-role.js [hunk 32]
```diff
@@ -0,0 +1,26 @@
...
+  async up(queryInterface, Sequelize) {
+    await queryInterface.createTable('Roles', {
+      id: {
+        allowNull: false,
+        autoIncrement: true,
+        primaryKey: true,
+        type: Sequelize.INTEGER,
+      },
+      name: {
+        type: Sequelize.ENUM('admin', 'requester', 'manager', 'super_admin'),
```
Comment (ericus123) [inline src/database/migrations/20220405214034-create-role.js]: @itisWasp  Use camelCase here.
change super_admin to **superAdmin**

File: src/swagger/roles.doc.js [hunk 33]
```diff
@@ -0,0 +1,74 @@
...
+
+  '/api/users': {
+    get: {
+      tags: ['Roles'],
+      description: 'Get all Registered Users',
+      parameters: [],
+      responses: {
+        201: {
+          description: 'Users Retrived Successfully',
+        },
```
Comment (ericus123) [inline src/swagger/roles.doc.js]: This is **200** status not 201

File: test/mocks/index.js [hunk 34]
```diff
@@ -65,7 +65,7 @@ export const dummyEmailUser = {
 
 export const realUser = {
   userName: 'kevinNew',
-  email: process.env.TEST_INIT_EMAIL,
+  email: 'demjinzwalvis2002@gmail.com',
```
Comment (ericus123) [inline test/mocks/index.js]: Why reveal emails like this ??

File: test/roles.test.js [hunk 35]
```diff
@@ -0,0 +1,77 @@
...
+    const response = await request(app).post('/api/auth/login').send({
+      email: process.env.ADMIN_ACCOUNT_EMAIL,
+      password: process.env.ADMIN_ACCOUNT_PASSWORD,
+    });
+
+    const Token = response.body.accessToken;
+    const res = await request(app)
+      .get(`/api/users`)
+      .set('Authorization', `Bearer ${Token}`);
+    expect(res.status).to.equal(201);
```
Comment (ericus123) [inline test/roles.test.js]: Again remember to change the status to 200 not 201

**Review by ericus123 — CHANGES_REQUESTED** [review]

File: src/database/migrations/20220417205059-update_roles_datatype.js [hunk 36]
```diff
@@ -0,0 +1,22 @@
+module.exports = {
+  async up(queryInterface, Sequelize) {
+    return Promise.all([
+      // queryInterface.addColumn('Roles', 'name', {
+      //   type: Sequelize.ENUM('admin', 'requester', 'manager', 'superAdmin'),
+      // }),
```
Comment (ericus123) [inline src/database/migrations/20220417205059-update_roles_datatype.js]: @itisWasp why is this commented ???
if not needed, should be removed

File: src/middlewares/superadminAuth.js [hunk 37]
```diff
@@ -0,0 +1,30 @@
...
+class superadminAuth {
+  static verifyToken = async (req, res, next) => {
+    const token = req.header('Authorization').replace('Bearer ', '');
+
+    if (!token) {
+      return res
+        .status(403)
+        .json({ message: 'Token Not Provided Please Check Headers' });
+    }
+
```
Comment (ericus123) [inline src/middlewares/superadminAuth.js]: @itisWasp  This is already implemented in another middleware.
Just call that middleware before this superAdminAuth in your route.
Before checking is someone is a superAdmin (role)  he must be authenticated first

**Review by ericus123 — CHANGES_REQUESTED** [review]

File: src/services/RolesServices.js [hunk 38]
```diff
@@ -0,0 +1,58 @@
...
+};
+export const FindRolesAndCount = (data) => {
+  const foundRoles = Role.findAndCountAll(data);
+  return foundRoles;
+};
+
+export const findRoleById = (query) => {
+  const role = Role.findOne({ where: { id: query.id } });
+  return role;
+};
```
Comment (ericus123) [inline src/services/RolesServices.js]: this is async

File: src/services/RolesServices.js [hunk 39]
```diff
@@ -0,0 +1,58 @@
...
+    {
+      where: { id: userId },
+    }
+  );
+  return changes;
+};
+export const deleteOne = (data) => {
+  const deleted = Role.destroy({ where: { id: data } });
+  return deleted;
+};
```
Comment (ericus123) [inline src/services/RolesServices.js]: this is async please

File: src/middlewares/superadminAuth.js [hunk 40]
```diff
@@ -0,0 +1,17 @@
+class superadminAuth {
+  static verifyToken = async (req, res, next) => {
```
Comment (ericus123) [inline src/middlewares/superadminAuth.js]: the naming here can be something else like checkAdmin/verifyAdmin but not just verifyToken

**Review by ericus123 — CHANGES_REQUESTED** [review]

File: src/services/RolesServices.js [hunk 41]
```diff
@@ -0,0 +1,58 @@
...
+import { Role } from '../database/models/role';
+
+export const createOne = async (data) => {
+  const created = await models.Role.create(data);
+  return created;
+};
+
+export const findRole = (data) => {
+  const found = Role.findOne({ where: { name: data.name } });
+  return found;
```
Comment (didasmbalanya) [inline src/services/RolesServices.js]: This is currently returning a promise, not a found Role.

Await the search

File: src/services/RolesServices.js (hunk 38 above)
Comment (didasmbalanya) [inline src/services/RolesServices.js]: returning a promise

File: src/controllers/user.js (hunk 29 above)
Comment (didasmbalanya) [reply src/controllers/user.js]: You can use the one in the roles service already,  "findRolesById", then add a second argument to the functions for other models to include

File: src/database/config/config.js (hunk 11 above)
Comment (didasmbalanya) [reply src/database/config/config.js]: please remove this

**Review by didasmbalanya — CHANGES_REQUESTED** [review]

File: src/database/models/role.js (hunk 30 above)
Comment (didasmbalanya) [reply src/database/models/role.js]: @Procech20 please learn what JSDOCs are and what they do for JS

**Review by ericus123 — APPROVED** [review]

File: .github/workflows/Node.js.yml [hunk 42]
```diff
@@ -56,3 +56,10 @@ jobs:
       HOST: ${{ secrets.HOST }}
       TEST_INIT_EMAIL: ${{ secrets.TEST_INIT_EMAIL }}
       TEST_REAL_EMAIL: ${{ secrets.TEST_REAL_EMAIL }}
+      TEST_REGISTER_EMAIL: ${{ secrets.TEST_REGISTER_EMAIL }}
+      ADMIN_ACCOUNT_FIRSTNAME: ${{secrets.ADMIN_ACCOUNT_FIRSTNAME}}
+      ADMIN_ACCOUNT_LASTNAME: ${{secrets.ADMIN_ACCOUNT_LASTNAME}}
+      ADMIN_ACCOUNT_USERNAME: ${{secrets.ADMIN_ACCOUNT_USERNAME}}
+      ADMIN_ACCOUNT_EMAIL: ${{secrets.ADMIN_ACCOUNT_EMAIL}}
+      ADMIN_ACCOUNT_PASSWORD: ${{secrets.ADMIN_ACCOUNT_PASSWORD}}
+      ADMIN_ACCOUNT_ROLEID: ${{secrets.ADMIN_ACCOUNT_ROLEID}}
```
Comment (niyojuste) [inline .github/workflows/Node.js.yml:65]: Add these variables to the .env.example as well

File: package.json [hunk 43]
```diff
@@ -9,15 +9,21 @@
...
+    "start": "yarn run build && cross-env NODE_ENV=production yarn migrate && NODE_ENV=production yarn create:admin && cross-env NODE_ENV=production node lib/index.js",
+    "dev": " yarn migrate && yarn create:admin && nodemon --exec babel-node -r dotenv/config src/index.js",
     "lint": "eslint src/** test/** --fix",
     "prepare": "husky install",
-    "test": "yarn migrate --env test && cross-env NODE_ENV=test c8 mocha --timeout 100000 --exit --require @babel/register"
+    "test": "yarn migrate --env test && cross-env NODE_ENV=test yarn create:admin && cross-env NODE_ENV=test c8 --reporter=text --reporter=lcov ./node_modules/.bin/mocha --timeout 200000 --exit --require @babel/register"
+  },
+  "nyc": {
+    "include": "src",
+    "exclude": "node_modules"
```
Comment (niyojuste) [inline package.json]: This is redudant since it is overwritten by the configuration in .nycrc.json, same as the reporters in the `test` script

File: test/mocks/index.js [hunk 44]
```diff
@@ -95,3 +95,39 @@ export const credentials = {
...
+};
+
+export const saveUser = {
+  firstName: registerUser.FirstName,
+  lastName: registerUser.LastName,
+  email: registerUser.Email,
+  userName: registerUser.UserName,
+  password: registerUser.Password,
+  roleId: registerUser.RoleId,
+};
```
Comment (niyojuste) [inline test/mocks/index.js]: Consider the **D**on't **R**epeat **Y**ourself approach

File: src/config/init.js [hunk 45]
```diff
@@ -0,0 +1,35 @@
...
+  });
+
+  if (!admin) {
+    createUser({
+      firstName: config.adminAccountFirstname,
+      lastName: config.adminAccountLastname,
+      email: config.adminAccountEmail,
+      userName: config.adminAccountUsername,
+      password: hashedPassword,
+      roleId: config.adminAccountRoleId,
```
Comment (niyojuste) [inline src/config/init.js]: Consider the **D**on't **R**epeat **Y**ourself approach

File: package.json [hunk 46]
```diff
@@ -9,15 +9,21 @@
...
+    "create:admin": "yarn run build && yarn create:roles && node lib/config/init.js",
     "build": "babel src -d lib",
-    "start": "yarn run build && cross-env NODE_ENV=production node lib/index.js",
-    "dev": "cross-env NODE_ENV=development nodemon --exec babel-node src/index.js",
+    "start": "yarn run build && cross-env NODE_ENV=production yarn migrate && NODE_ENV=production yarn create:admin && cross-env NODE_ENV=production node lib/index.js",
+    "dev": " yarn migrate && yarn create:admin && nodemon --exec babel-node -r dotenv/config src/index.js",
     "lint": "eslint src/** test/** --fix",
     "prepare": "husky install",
-    "test": "yarn migrate --env test && cross-env NODE_ENV=test c8 mocha --timeout 100000 --exit --require @babel/register"
+    "test": "yarn migrate --env test && cross-env NODE_ENV=test yarn create:admin && cross-env NODE_ENV=test c8 --reporter=text --reporter=lcov ./node_modules/.bin/mocha --timeout 200000 --exit --require @babel/register"
```
Comment (niyojuste) [inline package.json]: Calling the `create:admin` script is unnecessary since you're going to want to clear the database after each test suite, as you're already calling the function to create an admin when needed in the **test/roles.test.js** file

File: src/helpers/CheckUser.js [hunk 47]
```diff
@@ -0,0 +1,20 @@
...
+    .messages({
+      'string.pattern.base': '{{#label}} must contain a valid id',
+      'string.valid.base':
+        '{{#label}} must be written as 1, 2, 3 representing requester,manager, and admin respectively',
+    }),
+});
+
+export const checkUser = async (params) => {
+  return User.findOne({ where: params });
+};
```
Comment (niyojuste) [inline src/helpers/CheckUser.js]: There already is a service for this.

**Review by niyojuste — CHANGES_REQUESTED** [review]

File: src/database/models/user.js (hunk 16 above)
Comment (niyojuste) [reply src/database/models/user.js:10]: > good job here, just revert to es6 though

@itisWasp you seem to have missed this

File: test/mocks/index.js (hunk 44 above)
Comment (itisWasp) [reply test/mocks/index.js]: @niyojuste There is a reason for that please check thoroughly in my test you will see why I configured that.

File: src/config/init.js (hunk 45 above)
Comment (itisWasp) [reply src/config/init.js]: @niyojuste Check the flow please I am not Repeating myself.

File: package.json (hunk 46 above)
Comment (itisWasp) [reply package.json]: @niyojuste the admin must be created before the db is run regardless of my test other tests are using that.

**Review by Procech20 — APPROVED** [review]

File: test/roles.test.js [hunk 48]
```diff
@@ -0,0 +1,79 @@
...
+      await request(app).post('/api/auth/register').send(saveUser);
+
+      const User = await findIdUser(saveUser.email);
+      id = User.dataValues.id;
+
+      const res = await request(app).post('/api/auth/login').send({
+        email: saveUser.email,
+        password: saveUser.password,
+      });
+      console.log(res);
```
Comment (Procech20) [inline test/roles.test.js]: What is this console log doing here. Please remove it.

File: src/database/models/user.js [hunk 49]
```diff
@@ -1,14 +1,13 @@
-import { Model } from 'sequelize';
+const { Model } = require('sequelize');
 
-export default (sequelize, DataTypes) => {
+module.exports = (sequelize, DataTypes) => {
```
Comment (niyojuste) [inline src/database/models/user.js]: Change these back to ES6, please

File: src/helpers/CheckUser.js (hunk 47 above)
Comment (niyojuste) [inline src/helpers/CheckUser.js]: Please use the service already created. Although, I don't seem to find anywhere you used this function.

File: test/auth.test.js [hunk 50]
```diff
@@ -101,7 +101,4 @@ describe('Testing authentication routes', () => {
     expect(res.status).to.be.equal(200);
     expect(res.body).to.have.property('message', 'User logout successful');
   });
-  after(async () => {
-    await User.destroy({ where: {}, truncate: true });
-  });
```
Comment (niyojuste) [inline test/auth.test.js:106]: Why was this removed?

File: test/mocks/index.js (hunk 44 above)
Comment (niyojuste) [reply test/mocks/index.js]: Please look into this. You can have the values in **_registerUser_** be passed to **_saveUser_** instead. Also, notice you're referring to registerUser.roleId which is undefined. And, apart from being transferred to saveUser, registerUser doesn't seem to be used again, so better have on variable.

File: test/roles.test.js [hunk 51]
```diff
@@ -0,0 +1,79 @@
...
+let superToken;
+
+let id;
+
+describe('ROLE END-POINT TEST', async () => {
+  await adminScript();
+  describe('ASSIGNROLE USER TEST', () => {
+    before(async () => {
+      await request(app).post('/api/auth/register').send(saveUser);
+
```
Comment (niyojuste) [inline test/roles.test.js]: @itisWasp **Important Notice**
This test suite isn't being run, try removing `async` in the describe option and you can use a single describe in this case. Also, is there a particular reason you're not calling the `adminScript` in the before hook?

File: test/roles.test.js (hunk 48 above)
Comment (niyojuste) [inline test/roles.test.js]: I think you missed this `console.log`

**Review by niyojuste — CHANGES_REQUESTED** [review]

**Review by Robertishimwe — APPROVED** [review]
Looks good to me

**Review by Procech20 — APPROVED** [review]

File: package.json (hunk 43 above)
Comment (niyojuste) [inline package.json]: Remove this property along with the coverage `reporter` options in the `test` script. This is being handled in the .nycrc.json file (on the root of the project folder) with additional options that make the coverage fit our needs.

File: src/database/models/index.js [hunk 52]
```diff
@@ -25,11 +25,10 @@ fs.readdirSync(__dirname)
       file.indexOf('.') !== 0 && file !== basename && file.slice(-3) === '.js'
   )
   .forEach((file) => {
-    const model = require(path.join(__dirname, file)).default(
+    const model = require(path.join(__dirname, file))(
```
Comment (niyojuste) [inline src/database/models/index.js]: Add back this `default` function. Otherwise, the models written in ES6 would throw an error then, of course, change all the models to be written in ES6.

File: src/database/models/role.js [hunk 53]
```diff
@@ -0,0 +1,22 @@
+const { Model } = require('sequelize');
+
+module.exports = (sequelize, DataTypes) => {
```
Comment (niyojuste) [inline src/database/models/role.js]: Change this to ES6, please. Refer to previous comment on database/models/index.js

File: src/database/models/user.js (hunk 49 above)
Comment (niyojuste) [inline src/database/models/user.js]: Change this back to ES6, please. Refer to previous comment on database/models/index.js

**Review by niyojuste — CHANGES_REQUESTED** [review]

**Review by niyojuste — APPROVED** [review]
Nice!
